### PR TITLE
Fix build 32bit core build when  `CFG_VIRTUALIZATION=y

### PR DESCRIPTION
**Pull request by etienne-lms**
Fixing build of OP-TEE core when `CFG_ARM32_core=y` and  `CFG_VIRTUALIZATION=y`.
Note: `CFG_VIRTUALIZATION=y` still mandates `CFG_WITH_LPAE=y`.

Be aware that these 2 changes only allows building core without warning and error. These changes have not been tested.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/mm/core_mmu_lpae.c [hunk 1]
```diff
@@ -429,7 +429,11 @@ void core_mmu_set_prtn(struct mmu_partition *prtn)
 
 	ttbr = virt_to_phys(prtn->l1_tables[0][get_core_pos()]);
 
+#ifdef ARM32
+	write_ttbr0(ttbr | ((paddr_t)prtn->asid << (TTBR_ASID_SHIFT - 32)));
```
Comment (lorc) [inline core/arch/arm/mm/core_mmu_lpae.c]: `TTBR_ASID_SHIFT` is defined as `48` for both arm32 and arm64. Why do you need to subtract 32 there?

Comment (etienne-lms) [reply core/arch/arm/mm/core_mmu_lpae.c]: My fault. You're absolutely right. I mixed up 32b and 64b MMU.

**Comment by etienne-lms** [comment]
Please find corrected changes from review comment.

@lorc: I've added a commit regarding something that appeared inconsistent in the generic mapping (see 3rd commit). I need you feedback on this.

With these 3 changes, I observe equivalent behavior of OP-TEE for 64b and 32b modes when built with `CFG_VIRTUALIZATION=y`, that is OP-TEE boots then Linux successfully boots after OP-TEE but, without appropriate EL2 firmware, OP-TEE linux driver fails to probe with trace:
```
[    3.808242] optee: too small shared memory area
```

File: core/arch/arm/mm/core_mmu.c [hunk 2]
```diff
@@ -89,7 +89,7 @@ register_phys_mem_ul(MEM_AREA_TEE_RAM_RX, VCORE_UNPG_RX_PA, VCORE_UNPG_RX_SZ);
 register_phys_mem_ul(MEM_AREA_TEE_RAM_RO, VCORE_UNPG_RO_PA, VCORE_UNPG_RO_SZ);
 
 #ifdef CFG_VIRTUALIZATION
-register_phys_mem_ul(MEM_AREA_TEE_RAM_RO, VCORE_UNPG_RW_PA, VCORE_UNPG_RW_SZ);
+register_phys_mem_ul(MEM_AREA_TEE_RAM_RW, VCORE_UNPG_RW_PA, VCORE_UNPG_RW_SZ);
```
Comment (lorc) [inline core/arch/arm/mm/core_mmu.c]: This change is not  needed. What we are doing there is mapping of "golden" .data section. This memory will be then copied to .data section for new guest instance. So we don't want to make it read-write.

**Comment by lorc** [comment]
For "core: virt: fix format in memory map traces" and "core: mmu: fix MMU support with nsec virtualization for arm32":

`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`


And NAK for "core: fix mapping of unpaged rw section when CFG_VIRTUALIZATION=y": this is intended behaviour.

**Comment by etienne-lms** [comment]
> And NAK for "core: fix mapping of unpaged rw section when CFG_VIRTUALIZATION=y": this is intended behaviour.

Ok, thanks :)   I was pretty sure this was not so obvious.
Could you elaborate a bit why the read/write unpaged memory is mapped read-only? 
I think a nice inline comment in `core_mmu.c` would help not be confused.

I observed that 32bit core faults during its inits because of this mapping, while 64bit core does not. Maybe something still badly done in the 32bit case.

I've just added on more change i forgot to push previously: bypass call to `thread_clr_boot_thread()`.

**Comment by jforissier** [comment]
If the 32-bit virtualization is not tested I'd rather state this explicitly by rejecting `CFG_ARM32_core=y CFG_VIRTUALIZATION=y` in the main config file... doing otherwise is both useless and misleading IMO.

If however we have a working configuration (QEMU for instance), then it's another story.

File: core/arch/arm/mm/core_mmu.c (hunk 2 above)
Comment (etienne-lms) [reply core/arch/arm/mm/core_mmu.c]: @lorc and all, would you mind if we add an inline comment like `/* Map unpaged RW memory as RO, will be copied to .data for new guest instances */` ?

**Comment by etienne-lms** [comment]
@jforissier, i get you point, yet there is no reason why 32bit core cannot support virtual non-secure guests.

File: core/arch/arm/mm/core_mmu.c (hunk 2 above)
Comment (lorc) [reply core/arch/arm/mm/core_mmu.c]: Yes, that would be great. Actually, I wanted to suggest this.

**Comment by lorc** [comment]
> I observed that 32bit core faults during its inits because of this mapping, while 64bit core does not. Maybe something still badly done in the 32bit case.

This is exactly why it is mapped RO :)

Some part of TEE nexus is trying to access `.data` section for the per-guest data. I believe, some ARM32-specific variable is not marked with `__nex_data` attribute.

**Comment by lorc** [comment]
@jforissier,

I agree with you, that it is should be explicitly disabled. On other hand, I welcome any patches that at least partially fix the arm32 issues.

**Pull request closed without merging**